pkg/agent/trafficfilter/traffic_filter_linux.go

```diff
@@ -1,4 +1,4 @@
-// Copyright 2019 Antrea Authors
```

Comment:
I thought we could still name it iptables.go, and even keep it in package iptables, but implement trafficfilter Client interface. What you think?

Reply:
Personally, I perfer trafficfilter, as `iptables` is a name strongly bound to Linux, and `trafficfilter` sounds more generic.  What's your thoughts @tnqn ?

Reply:
I was actually thinking about adjusting these interfaces too but with larger scope, as currently these route and iptables interfaces sounds really implementation oriented and deeply coupled with each other (e.g. we need to use a specific mark in iptables then match the mark in routes, the routes cannot work without iptables allowing forwarding) while they are just sub-processes for same purpose: making the right route for container traffic. 
So I'm thinking we don't really need these interfaces to call `c.routeClient.AddPeerCIDRRoute` and `c.iptablesClient.AddPeerCIDR` in NodeRouteController separately, there should be just one purpose facing interface like RouteService whose implementations holds routeClient, iptablesClient on linux and netroute.Route on windows.
```
# interfaces.go
type RouteService interface {
	// AddRoute adds a route to the provided podCIDR.
	AddRoute(podCIDR *net.IPNet, linkIndex int, peerNodeIP, peerGwIP net.IP)

	// DeleteRoute deletes a route to the provided podCIDR.
	DeleteRoute(podCIDR *net.IPNet)
}
```
We can benefit these:
1. It frees its consumers AgentInitializer and NodeRouteController from dealing with multiple interfaces for same purpose.
2. It frees developer from figuring out generic interface and function names for OS specific logic
3. It encapsulates the routing logic (except openflow) in a single interface, making it easier to change the implementation without changing consumer code, like I want to propose changing the way of matching podCIDRs in iptables to using ipset for performance and ease of management (currently it doesn't even clean up podCIDR iptables rule when a Node leaves cluster), it's easier to add an ipset operation to the implementation without concerning it's not suitable be in an iptables interface if in old way.

I can have a quick PR to express the proposal if this makes sense to you, then this PR can focus on providing a windows implementation of RouteService, what do you think? @jianjuns @wenyingd

Reply:
I thought about this idea too (moving iptables call to routeClient), but not very sure iptables logics should belong to routeClient or not - besides AddPeerCIDR, we also have iptables initialize() and reconcile(). Where you are going to place them?

Back to the original topic of this thread - without the new refactoring you are proposing, do you think we should still keep iptables package and iptables.go even it implements another interface? I think so, as the code is really about iptables.

Reply:
The interface will provide methods `Initialize(nodeConfig *config.NodeConfig)` and `Reconcile(podCIDRs []*net.IPNet)` too, and even currently what the two clients actually required are exactly the same.

Without refactoring, I agree with your point, kube-proxy's multiple implementation are putting them in separate packets named by their implementation, I think the names are informative. 
See https://github.com/kubernetes/kubernetes/tree/master/pkg/proxy. 
`types.go` is the interface.
package `ipvs` implements the interface `Provider` via ipvs focused mechanism.
package `iptables` implements it via iptables focused mechanism.

If keeping iptables/iptables.go, how are we going to organize the interface and the iptables implementation, it looks like too many packages under pkg/agent/: pkg/agent/trafficfilter, /pkg/agent/iptables and another package for windows implementation, and hard to know their relationship from package structure.

Reply:
I meant iptables.initialize() also add rules/tables not directly relevant to routes? If not, I would quite support your proposal.

If it is better, we can put iptables.go under the trafficfilter package, or a sub-package?